File: agents/evaluation_agent.py
from agents.state import GraphState


class EvaluationAgentNode:
    """LangGraph node that defines evaluation criteria."""

    def __call__(self, state: GraphState) -> dict:
        # 1. Retrieve data from State
        diagnosis = state["diagnosis"]
        concept = diagnosis.get("weakest_concept")

        # 2. Build evaluation plan
        evaluation_plan = {
            "concept": concept,
            "success_signals": [
                "fewer repeated mistakes on similar problems",
                "lower time-to-solve",
                "higher confidence on the concept",
                "ability to solve one fresh problem without hints",
            ],
            "recheck_after": "next focused practice block or mini-checkpoint",
            "replan_trigger": "if the same error pattern appears again after the intervention",
        }

        # 3. Return the key updates to the GraphState
        return {
            "evaluation_plan": evaluation_plan,
            "trace": state["trace"] + [{"agent": "evaluation-agent", "message": "Defined multi-signal evaluation criteria and replanning trigger."}],
        }

File: backend-node/src/agents/evaluationAgent.js
import { invokeAgentStructured } from "../services/llmClient.js";
import { EVALUATOR_PROMPT } from "./prompts.js";
import { EvaluationOutputSchema } from "./agentSchemas.js";

/**
 * Defines evaluation criteria using an LLM.
 */
export async function evaluate(state) {
  const { diagnosis, knowledge_base: kb } = state;

  const concept = diagnosis.weakest_concept;
  const misconception = diagnosis.misconception;

  const context = {
    diagnosed_concept: concept,
    misconception,
    evaluation_rules: (kb.evaluation_rules?.rules || []).filter((r) => r.concept_id === concept),
  };

  const userContent = `Context:\n${JSON.stringify(context, null, 2)}`;

  const result = await invokeAgentStructured({
    systemPrompt: EVALUATOR_PROMPT,
    userContent,
    responseSchema: EvaluationOutputSchema,
  });

  state.evaluation_plan = {
    concept,
    success_signals: result.success_signals,
    recheck_after: result.recheck_after,
    replan_trigger: result.replan_trigger,
    mastery_threshold: result.mastery_threshold,
  };

  state.trace.push({
    agent: "evaluation-agent",
    message: "Defined multi-signal evaluation criteria via LLM.",
  });

  return state;
}
